lib/turn_endpoint.ex

```diff
@@ -0,0 +1,624 @@
+  use [membrane_funnel_plugin](https://github.com/membraneframework/membrane_funnel_plugin).
+
+  ### Messages API
+  You can send following messages to TURN Endpoint:
+
+  - `:gather_candidates`
+
+  - `{:set_remote_credentials, credentials}` - credentials are string in form of "ufrag passwd"
+
+  - `:peer_candidate_gathering_done`
+
+  ### Notifications API
+  - `{:new_candidate_full, candidate}`
+    Triggered by: `:gather_candidates`
+
+  - `{:integrated_turn_servers, integrated_turn_servers}`
+
+  - `{:handshake_init_data, component_id, handshake_init_data}`
+
+  - `{:connection_ready, stream_id, component_id}`
+
+  - `{:component_state_failed, @stream_id, @component_id}`
+
+  ### Sending and receiving messages
+  To send or receive messages just link to TURN Endpoint using relevant pads.
+  As soon as connection is established your element will receive demands and incoming messages.
+  """
+  use Membrane.Filter
+
+  alias Membrane.TURN.{Utils, Handshake}
+  alias Membrane.Funnel
+
+  require Membrane.Logger
+
+  @component_id 1
+  @stream_id 1
+  @fake_candidate_addr {{127, 0, 0, 1}, 41847}
+
+  @typedoc """
+  Options defining the behavior of TURN.Endpoint in relation to integrated TURN servers.
+  - `:ip` - IP, where integrated TURN server will open its sockets
+  - `:mock_ip` - IP, that will be put part of XOR-RELAYED-ADDRESS attribute in Allocation Succes message.
+  Might be, but doesn't have to be equal to `:ip`
+  - `:ports_range` range, where integrated TURN server will try to open ports
+  """
+  @type integrated_turn_options_t() :: [
+          use_integrated_turn: boolean(),
+          ip: :inet.ip4_address() | nil,
+          mock_ip: :inet.ip4_address() | nil,
+          ports_range: {:inet.port_number(), :inet.port_number()} | nil
+        ]
+
+  def_options handshake_module: [
+                spec: module(),
+                default: Handshake.Default,
+                description: "Module implementing Handshake behaviour"
+              ],
+              handshake_opts: [
+                spec: keyword(),
+                default: [],
+                description:
+                  "Options for handshake module. They will be passed to init function of hsk_module"
+              ],
+              integrated_turn_options: [
+                spec: [integrated_turn_options_t()],
+                default: %{use_integrated_turn: false},
+                description: "Integrated TURN Options"
+              ]
+
+  def_input_pad :input,
+    availability: :on_request,
+    caps: :any,
+    mode: :pull,
+    demand_unit: :buffers
+
+  def_output_pad :output,
+    availability: :on_request,
+    caps: :any,
+    mode: :push
+
+  @impl true
+  def handle_init(options) do
+    %__MODULE__{
+      integrated_turn_options: integrated_turn_options,
+      handshake_module: hsk_module,
+      handshake_opts: hsk_opts
+    } = options
+
+    integrated_turn_servers = start_integrated_turn_servers(integrated_turn_options, self())
```

Comment:
It might be prettier, to put this in `handle_prepared_to_playing`, but, on the other hand, there is a chance, that parent of `TURN.Endpoint` might want to receive notification `{:integrated_turn_servers, ...}` as fast, as it is possible (this case occurs in `RTC.Engine`)